File: HangmanReconstructedV4.py
import random
import os

WELCOME = "Welcome to HANGMAN!"
ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
GAME_ABANDONED = 1
GAME_LOST = 2
GAME_WON = 3
MAXWRONG = 6

TOO_LONG = "Invalid: You must enter only 1 letter at a time!"
NOT_A_LETTER = "Invalid: The character you typed was not a letter!"
DUPLICATE_WRONG = "Invalid: You have already guessed this letter and it was wrong!"
DUPLICATE_RIGHT = "Invalid: You have already guessed this letter and it was correct!"
LETTER_WRONG = "Sorry, the word does not contain this letter."
LETTER_RIGHT = "Well done! A correct letter!"

DEFAULT_WORDS = [
    "AUTOMOBILE", "NETWORKING", "PRACTICAL",
    "CONGRESS", "COMMANDER", "STAPLER", "ENTERPRISE",
    "ESCALATION", "HAPPINESS", "WEDNESDAY", "THUNDER",
    "MARATHON", "LABORATORY", "HARBINGER", "SUNSHINE",
    "JOURNEY", "FANTASTIC", "DISCOVERY", "BOOKCASE",
    "HANGMAN"
]

class HangmanGame:
    def __init__(self):
        self.game_word = self.pick_random_word()
        self.guess_tracker = [0] * 26
        self.partial_solution = "_" * len(self.game_word)
        self.no_misses = 0
        self.game_ended = False
        self.outcome = None
    
    def pick_random_word(self):
        word_list_file = "WordList.txt"
        try:
            if os.path.exists(word_list_file):
                with open(word_list_file, 'r') as f:
                    words = [w.strip().upper() for w in f.readlines() if w.strip()]
                    if words:
                        return random.choice(words)
        except IOError:
            pass
        return random.choice(DEFAULT_WORDS)
    
    def build_hangman(self, misses):
        stages = [
            "   +---+\n   |   |\n       |\n       |\n       |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n       |\n       |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n   |   |\n       |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n  /|   |\n       |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n  /|\\  |\n       |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n  /|\\  |\n  /    |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n  /|\\  |\n  / \\  |\n       |\n  ======="
        ]
        clamped_misses = max(0, min(6, misses))
        return stages[clamped_misses]
    
    def get_wrong_guesses(self):
        wrong_guesses = ""
        for i in range(26):
            if self.guess_tracker[i] == -1:
                wrong_guesses += ALPHABET[i] + " "
        return wrong_guesses.strip() if wrong_guesses.strip() else "None"
    
    def get_choice(self, message):
        wrong_guesses = self.get_wrong_guesses()
        prompt = f"\n\n{message}\n\n\tWord: {self.partial_solution}\n\tMisses: {self.no_misses} / {MAXWRONG}\n\tIncorrect: {wrong_guesses}\n\tHangman:\n{self.build_hangman(self.no_misses)}\n\nType one letter, or leave blank to abandon the game.\n> "
        return input(prompt)
    
    def play(self):
        choice = self.get_choice(WELCOME)
        
        while not self.game_ended:
            if choice == "":
                self.game_ended = True
                self.outcome = GAME_ABANDONED
            elif len(choice) > 1:
                choice = self.get_choice(TOO_LONG)
            else:
                choice = choice.upper()
                letter_pos = ALPHABET.find(choice)
                
                if letter_pos == -1:
                    choice = self.get_choice(NOT_A_LETTER)
                elif self.guess_tracker[letter_pos] == -1:
                    choice = self.get_choice(DUPLICATE_WRONG)
                elif self.guess_tracker[letter_pos] == 1:
                    choice = self.get_choice(DUPLICATE_RIGHT)
                else:
                    game_letter_pos = self.game_word.find(choice)
                    
                    if game_letter_pos == -1:
                        self.no_misses += 1
                        self.guess_tracker[letter_pos] = -1
                        
                        if self.no_misses >= MAXWRONG:
                            self.game_ended = True
                            self.outcome = GAME_LOST
                        else:
                            choice = self.get_choice(LETTER_WRONG)
                    else:
                        new_partial_sol = ""
                        for i in range(len(self.game_word)):
                            if self.game_word[i] == choice:
                                new_partial_sol += choice
                            else:
                                new_partial_sol += self.partial_solution[i]
                        
                        self.partial_solution = new_partial_sol
                        self.guess_tracker[letter_pos] = 1
                        
                        if self.partial_solution == self.game_word:
                            self.game_ended = True
                            self.outcome = GAME_WON
                        else:
                            choice = self.get_choice(LETTER_RIGHT)
    
    def report_outcome(self):
        if self.outcome == GAME_ABANDONED:
            print(f"You abandoned the game. The word was {self.game_word}")
        elif self.outcome == GAME_LOST:
            print(f"{LETTER_WRONG}\nYou lose. The word was {self.game_word}")
        elif self.outcome == GAME_WON:
            print(f"{LETTER_RIGHT}\nCongratulations! You win! The word was {self.game_word}")

if __name__ == "__main__":
    game = HangmanGame()
    game.play()
    game.report_outcome()

File: HangmanReconstructedV4.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Random;

public class HangmanReconstructedV4 {
    static final String WELCOME = "Welcome to HANGMAN!";
    static final String ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    static final int GAME_ABANDONED = 1;
    static final int GAME_LOST = 2;
    static final int GAME_WON = 3;
    static final int MAXWRONG = 6;
    
    static final String TOO_LONG = "Invalid: You must enter only 1 letter at a time!";
    static final String NOT_A_LETTER = "Invalid: The character you typed was not a letter!";
    static final String DUPLICATE_WRONG = "Invalid: You have already guessed this letter and it was wrong!";
    static final String DUPLICATE_RIGHT = "Invalid: You have already guessed this letter and it was correct!";
    static final String LETTER_WRONG = "Sorry, the word does not contain this letter.";
    static final String LETTER_RIGHT = "Well done! A correct letter!";
    
    static final String[] DEFAULT_WORDS = {
        "AUTOMOBILE", "NETWORKING", "PRACTICAL",
        "CONGRESS", "COMMANDER", "STAPLER", "ENTERPRISE",
        "ESCALATION", "HAPPINESS", "WEDNESDAY", "THUNDER",
        "MARATHON", "LABORATORY", "HARBINGER", "SUNSHINE",
        "JOURNEY", "FANTASTIC", "DISCOVERY", "BOOKCASE",
        "HANGMAN"
    };
    
    String gameWord;
    int[] guessTracker;
    String partialSolution;
    int noMisses;
    boolean gameEnded;
    int outcome;
    
    BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    Random random = new Random();
    
    public HangmanReconstructedV4() {
        gameWord = pickRandomWord();
        guessTracker = new int[26];
        partialSolution = "_".repeat(gameWord.length());
        noMisses = 0;
        gameEnded = false;
        outcome = 0;
    }
    
    String pickRandomWord() {
        File wordListFile = new File("WordList.txt");
        try {
            if (wordListFile.exists()) {
                ArrayList<String> words = new ArrayList<>();
                BufferedReader fileReader = new BufferedReader(new FileReader(wordListFile));
                String line;
                while ((line = fileReader.readLine()) != null) {
                    String word = line.trim().toUpperCase();
                    if (!word.isEmpty()) {
                        words.add(word);
                    }
                }
                fileReader.close();
                if (!words.isEmpty()) {
                    return words.get(random.nextInt(words.size()));
                }
            }
        } catch (IOException e) {
            // Fall through to defaults
        }
        return DEFAULT_WORDS[random.nextInt(DEFAULT_WORDS.length)];
    }
    
    String buildHangman(int misses) {
        String[] stages = {
            "   +---+\n   |   |\n       |\n       |\n       |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n       |\n       |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n   |   |\n       |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n  /|   |\n       |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n  /|\\  |\n       |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n  /|\\  |\n  /    |\n       |\n  =======",
            "   +---+\n   |   |\n   O   |\n  /|\\  |\n  / \\  |\n       |\n  ======="
        };
        int clampedMisses = Math.max(0, Math.min(6, misses));
        return stages[clampedMisses];
    }
    
    String getWrongGuesses() {
        StringBuilder wrongGuesses = new StringBuilder();
        for (int i = 0; i < 26; i++) {
            if (guessTracker[i] == -1) {
                wrongGuesses.append(ALPHABET.charAt(i)).append(" ");
            }
        }
        String result = wrongGuesses.toString().trim();
        return result.isEmpty() ? "None" : result;
    }
    
    String getChoice(String message) throws IOException {
        String wrongGuesses = getWrongGuesses();
        String prompt = String.format("\n\n%s\n\n\tWord: %s\n\tMisses: %d / %d\n\tIncorrect: %s\n\tHangman:\n%s\n\nType one letter, or leave blank to abandon the game.\n> ",
            message, partialSolution, noMisses, MAXWRONG, wrongGuesses, buildHangman(noMisses));
        System.out.print(prompt);
        return reader.readLine();
    }
    
    void play() throws IOException {
        String choice = getChoice(WELCOME);
        
        while (!gameEnded) {
            if (choice.isEmpty()) {
                gameEnded = true;
                outcome = GAME_ABANDONED;
            } else if (choice.length() > 1) {
                choice = getChoice(TOO_LONG);
            } else {
                choice = choice.toUpperCase();
                int letterPos = ALPHABET.indexOf(choice);
                
                if (letterPos == -1) {
                    choice = getChoice(NOT_A_LETTER);
                } else if (guessTracker[letterPos] == -1) {
                    choice = getChoice(DUPLICATE_WRONG);
                } else if (guessTracker[letterPos] == 1) {
                    choice = getChoice(DUPLICATE_RIGHT);
                } else {
                    int gameLetterPos = gameWord.indexOf(choice);
                    
                    if (gameLetterPos == -1) {
                        noMisses++;
                        guessTracker[letterPos] = -1;
                        
                        if (noMisses >= MAXWRONG) {
                            gameEnded = true;
                            outcome = GAME_LOST;
                        } else {
                            choice = getChoice(LETTER_WRONG);
                        }
                    } else {
                        StringBuilder newPartialSol = new StringBuilder();
                        for (int i = 0; i < gameWord.length(); i++) {
                            if (gameWord.charAt(i) == choice.charAt(0)) {
                                newPartialSol.append(choice);
                            } else {
                                newPartialSol.append(partialSolution.charAt(i));
                            }
                        }
                        
                        partialSolution = newPartialSol.toString();
                        guessTracker[letterPos] = 1;
                        
                        if (partialSolution.equals(gameWord)) {
                            gameEnded = true;
                            outcome = GAME_WON;
                        } else {
                            choice = getChoice(LETTER_RIGHT);
                        }
                    }
                }
            }
        }
    }
    
    void reportOutcome() {
        switch (outcome) {
            case GAME_ABANDONED:
                System.out.println("You abandoned the game. The word was " + gameWord);
                break;
            case GAME_LOST:
                System.out.println(LETTER_WRONG + "\nYou lose. The word was " + gameWord);
                break;
            case GAME_WON:
                System.out.println(LETTER_RIGHT + "\nCongratulations! You win! The word was " + gameWord);
                break;
        }
    }
    
    public static void main(String[] args) throws IOException {
        HangmanReconstructedV4 game = new HangmanReconstructedV4();
        game.play();
        game.reportOutcome();
    }
}
